src/humancellatlas/data/metadata/helpers/staging_area.py

```diff
@@ -0,0 +1,323 @@
+        def parse_file_name(file_name: str) -> Sequence[str]:
+            suffix = '.json'
+            assert file_name.endswith(suffix), file_name
+            return file_name[:-len(suffix)].split('_')
+
+        schema_type = content['schema_type']
+        if schema_type == 'links':
+            subgraph_id, version, project_id = parse_file_name(file_name)
+            return LinksFile(uuid=subgraph_id,
+                             version=version,
+                             name=file_name,
+                             content=content,
+                             project_id=project_id)
+        else:
+            entity_id, version = parse_file_name(file_name)
+            if schema_type == 'file_descriptor':
+                return DescriptorFile(uuid=entity_id,
+                                      version=version,
+                                      name=file_name,
+                                      content=content)
+            else:  # 'biomaterial', 'protocol', 'file', ...
+                return MetadataFile(uuid=entity_id,
+                                    version=version,
+                                    name=file_name,
+                                    content=content)
+
+
+@attr.s(frozen=True, kw_only=True, auto_attribs=True)
+class LinksFile(JsonFile):
+    """
+    A file describing the links between entities in a subgraph.
+    """
+    project_id: str
+
+
+@attr.s(frozen=True, kw_only=True, auto_attribs=True)
+class MetadataFile(JsonFile):
+    """
+    A file describing one entity (e.g. biomaterial, protocol) in a subgraph.
+    """
+    pass
+
+
+@attr.s(frozen=True, kw_only=True, auto_attribs=True)
+class DescriptorFile(JsonFile):
+    """
+    A file containing the checksums and other information for asserting the
+    integrity of a data file.
+    """
+    namespace: ClassVar[UUID] = UUID('5767014a-c431-4019-8703-0ab1b3e9e4d0')
+
+    @property
+    def to_manifest_entry(self):
+        """
+        The content of a descriptor transformed into a format ready to create a
+        ManifestEntry object.
+        """
+        return {
+            'content-type': self.content['content_type'],
+            'crc32c': self.content['crc32c'],
+            'indexed': False,
+            'name': self.content['file_name'],
+            's3_etag': self.content['s3_etag'],
+            'sha1': self.content['sha1'],
+            'sha256': self.content['sha256'],
+            'size': self.content['size'],
+            'uuid': str(uuid5(self.namespace, self.content['file_name'])),
+            'version': self.content['file_version']
+        }
+
+
+@attr.s(frozen=True, kw_only=True, auto_attribs=True)
+class StagingArea:
+    links: Mapping[str, LinksFile]  # Key is the subgraph ID aka links_id
+    metadata: Mapping[str, MetadataFile]  # Key is the entity ID
+    descriptors: Mapping[str, DescriptorFile]  # Key is the entity ID
+
+    def entity_ids_by_type(self,
```

Comment:
This looks like it should be protected. And methods that make up the public interface of a class should be listed first.